src/usb.rs

```diff
@@ -1,39 +1,54 @@
-//! Support code for working with the USB peripheral
-//!
-//! This module is different from the other peripheral APIs, in the sense that
-//! it isn't really a peripheral API, just some support code to make working
-//! with the USB peripheral possible. This is due to the existing STM32
-//! ecosystem support for USB.
+//! Interface to the USB peripheral
 //!
 //! As USB seems to work mostly the same across all STM32 MCUs, there is a
-//! single crates that supports USB for these MCUs:
+//! single crate that supports USB for these MCUs:
 //! [`stm32-usbd`](https://crates.io/crates/stm32-usbd)
 //!
-//! Since `stm32-usbd` depends on the HAL libraries for each STM32 family it
-//! supports, we can't exactly depend on it to provide everything you need here.
-//! Instead, we just provide some code that takes care of the platform-specific
-//! bits, which you can call before using `stm32-usbd` for the rest.
+//! This module implements some bits needed for `stm32-usbd` to work and exports
+//! `UsbBus` from `stm32-usbd`.
 //!
 //! Please check out the USB examples in the `examples/` directory to see how it
 //! fits together.
 
+use crate::stm32::{RCC, USB};
+use stm32_usbd::UsbPeripheral;
+
+use crate::gpio::gpioa::{PA11, PA12};
+use crate::gpio::{Floating, Input};
+pub use stm32_usbd::UsbBus;
+
+pub struct Peripheral {
+    pub usb: USB,
+    pub pin_dm: PA11<Input<Floating>>,
+    pub pin_dp: PA12<Input<Floating>>,
+}
 
-use crate::rcc::{
-    HSI48,
-    Rcc,
-};
+unsafe impl Sync for Peripheral {}
 
+unsafe impl UsbPeripheral for Peripheral {
+    const REGISTERS: *const () = USB::ptr() as *const ();
+    const DP_PULL_UP_FEATURE: bool = true;
+    const EP_MEMORY: *const () = 0x4000_6000 as _;
+    const EP_MEMORY_SIZE: usize = 1024;
 
-/// Initializes the USB peripheral
-///
-/// This method takes care of the platform-specific bits of the USB
-/// initialization. After calling this method, you need `stm32-usbd` to actually
-/// do anything useful with the USB peripheral.
-pub fn init(rcc: &mut Rcc, _: HSI48) {
-    // Reset USB peripheral
-    rcc.rb.apb1rstr.modify(|_, w| w.usbrst().set_bit());
-    rcc.rb.apb1rstr.modify(|_, w| w.usbrst().clear_bit());
+    fn enable() {
+        let rcc = unsafe { (&*RCC::ptr()) };
 
-    // We don't need to enable the USB peripheral, as stm32-usbd takes care of
-    // that.
+        cortex_m::interrupt::free(|_| {
+            // Enable USB peripheral
+            rcc.apb1enr.modify(|_, w| w.usben().set_bit());
+
+            // Reset USB peripheral
+            rcc.apb1rstr.modify(|_, w| w.usbrst().set_bit());
+            rcc.apb1rstr.modify(|_, w| w.usbrst().clear_bit());
+        });
+    }
+
+    fn startup_delay() {
+        // There is a chip specific startup delay. For STM32F103xx it's 1µs and this should wait for
```

Comment:
```suggestion
        // There is a chip specific startup delay. For STM32L0x2/x3 it's 1µs and this should wait for
```

This seems to be true for all STM32L0x2/x3, according to their data sheets.